tests/ansi_code_markup_tests.cpp

```diff
@@ -0,0 +1,253 @@
+    EXPECT_EQ(convert_ansi_markup(str), "\033[31;1mthis colour is red");
+}
+
+TEST(ConvertAnsiMarkup, InvalidForegroundColorEnd)
+{
+    const char str[] = "the color will be reset[/color]";
+    EXPECT_EQ(convert_ansi_markup(str), "the color will be reset[/color]");
+}
+
+TEST(ConvertAnsiMarkup, ValidForegroundLightBlue)
+{
+    const char str[] = "this [color=light-blue]colour is light blue";
+    EXPECT_EQ(convert_ansi_markup(str), "this \033[34mcolour is light blue");
+}
+
+TEST(ConvertAnsiMarkup, InvalidForegroundColor)
+{
+    const char str[] = "[color=invalid]this is an invalid foreground color";
+    EXPECT_EQ(convert_ansi_markup(str), "[color=invalid]this is an invalid foreground color");
+}
+
+TEST(ConvertAnsiMarkup, ValidBackgroundColorMiddle)
+{
+    const char str[] = "this [bgcolor=red]colour is red";
+    EXPECT_EQ(convert_ansi_markup(str), "this \033[41;1mcolour is red");
+}
+
+TEST(ConvertAnsiMarkup, ValidBackgroundColorStart)
+{
+    const char str[] = "[bgcolor=red]this colour is red";
+    EXPECT_EQ(convert_ansi_markup(str), "\033[41;1mthis colour is red");
+}
+
+TEST(ConvertAnsiMarkup, InvalidBackgroundColorEnd)
+{
+    const char str[] = "the color will be reset[/bgcolor]";
+    EXPECT_EQ(convert_ansi_markup(str), "the color will be reset[/bgcolor]");
+}
+
+TEST(ConvertAnsiMarkup, ValidBackgroundLightBlue)
+{
+    const char str[] = "this [bgcolor=light-blue]colour is light blue";
+    EXPECT_EQ(convert_ansi_markup(str), "this \033[44mcolour is light blue");
+}
+
+TEST(ConvertAnsiMarkup, InvalidBackgroundColor)
+{
+    const char str[] = "[bgcolor=invalid]this is an invalid foreground color";
+    EXPECT_EQ(convert_ansi_markup(str), "[bgcolor=invalid]this is an invalid foreground color");
+}
+
+TEST(ConvertAnsiMarkup, ValidForegroundColorMultiple)
+{
+    const char str[] = "this [color=red]colour is red. [color=blue]And this is blue.[reset]";
+    EXPECT_EQ(convert_ansi_markup(str), "this \033[31;1mcolour is red. \033[34;1mAnd this is blue.\033[0m");
+}
+
+TEST(ConvertAnsiMarkup, InvalidForegroundColorNoValue)
+{
+    const char str[] = "this [color]colour is red.";
+    EXPECT_EQ(convert_ansi_markup(str), "this [color]colour is red.");
+}
+
+TEST(ConvertAnsiMarkup, InvalidBackgroundColorNoValue)
+{
+    const char str[] = "this [bgcolor]colour is red.";
+    EXPECT_EQ(convert_ansi_markup(str), "this [bgcolor]colour is red.");
+}
+
+TEST(ConvertAnsiMarkup, InvalidTagName)
+{
+    const char str[] = "this [sometag] tag is invalid.";
+    EXPECT_EQ(convert_ansi_markup(str), "this [sometag] tag is invalid.");
+}
+
+TEST(ConvertAnsiMarkup, InvalidTagUnclosed)
+{
+    const char str[] = "this [color=red colour is red.";
+    EXPECT_EQ(convert_ansi_markup(str), "this [color=red colour is red.");
+}
+
+TEST(ConvertAnsiMarkup, InvalidTagUnopened)
+{
+    const char str[] = "this color=red] colour is red.";
+    EXPECT_EQ(convert_ansi_markup(str), "this color=red] colour is red.");
+}
+
```

Comment:
Great invalid coverage - can't see much else!
Maybe valid but nested tags (for an overall invalid use-case?)
```[bgcolor=light-blue [bgcolor=light-blue] ]```

Reply:
I think quotes and escapes are common in the comments - maybe pepper some of those into the invalid cases?
`\[bgcolor=light-blue\]`
`[bgcolor="light-blue]`
`["bgcolor='light-blue]`
... 
(for what its worth.. )

Reply:
> Great invalid coverage - can't see much else! Maybe valid but nested tags (for an overall invalid use-case?) `[bgcolor=light-blue [bgcolor=light-blue] ]`

Well, in this specific case, the nested tag should work, but not the outer. I haven't implemented a full-blown state-tracking parser here...

Reply:
Yup; that makes sense - and didn't mean to imply that the nesting should work (just that it shouldn't crash, and should ignore the outer tag.. or more specifically, treat it as ordinary unmatched text)